cmd/quadlet/main.go

```diff
@@ -103,28 +108,92 @@ func Debugf(format string, a ...interface{}) {
 func getUnitDirs(rootless bool) []string {
 	// Allow overriding source dir, this is mainly for the CI tests
 	unitDirsEnv := os.Getenv("QUADLET_UNIT_DIRS")
+	dirs := make([]string, 0)
+
 	if len(unitDirsEnv) > 0 {
-		return strings.Split(unitDirsEnv, ":")
+		for _, eachUnitDir := range strings.Split(unitDirsEnv, ":") {
+			if !filepath.IsAbs(eachUnitDir) {
+				Logf("%s not a valid file path", eachUnitDir)
+				return nil
+			}
+			dirs = appendSubPaths(dirs, eachUnitDir, false)
+		}
+		return dirs
 	}
 
-	dirs := make([]string, 0)
 	if rootless {
 		configDir, err := os.UserConfigDir()
 		if err != nil {
 			fmt.Fprintf(os.Stderr, "Warning: %v", err)
 			return nil
 		}
-		dirs = append(dirs, path.Join(configDir, "containers/systemd"))
+		dirs = appendSubPaths(dirs, path.Join(configDir, "containers/systemd"), false)
 		u, err := user.Current()
 		if err == nil {
-			dirs = append(dirs, filepath.Join(quadlet.UnitDirAdmin, "users", u.Uid))
+			dirs = appendNonNumericSubPaths(dirs, filepath.Join(quadlet.UnitDirAdmin, "users"))
+			dirs = appendSubPaths(dirs, filepath.Join(quadlet.UnitDirAdmin, "users", u.Uid), true)
 		} else {
 			fmt.Fprintf(os.Stderr, "Warning: %v", err)
 		}
 		return append(dirs, filepath.Join(quadlet.UnitDirAdmin, "users"))
 	}
-	dirs = append(dirs, quadlet.UnitDirAdmin)
-	return append(dirs, quadlet.UnitDirDistro)
+
+	dirs = appendSubPaths(dirs, quadlet.UnitDirAdmin, false)
+	return appendSubPaths(dirs, quadlet.UnitDirDistro, false)
+}
+
+func appendSubPaths(dirs []string, path string, isUserFlag bool) []string {
+	err := filepath.Walk(path, func(_path string, info os.FileInfo, err error) error {
+		if info == nil || info.IsDir() {
+			if strings.Contains(_path, filepath.Join(quadlet.UnitDirAdmin, "users")) {
+				if isUserFlag {
+					dirs = append(dirs, _path)
+				}
+			} else {
+				dirs = append(dirs, _path)
+			}
+		}
+		return err
+	})
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			Debugf("Error occurred walking sub directories %q: %s", path, err)
+		}
+	}
+	return dirs
+}
+
+func appendNonNumericSubPaths(dirs []string, path string) []string {
+	err := filepath.Walk(path, func(_path string, info os.FileInfo, err error) error {
+		if info == nil || info.IsDir() {
+			if strings.Contains(_path, filepath.Join(quadlet.UnitDirAdmin, "users")) {
+				if SystemUserDirsExist(_path) {
+					if !govalidator.IsNumeric(SystemUserLevelDirNameGet(_path)) {
+						dirs = append(dirs, _path)
+					}
+				}
+			} else {
+				dirs = append(dirs, _path)
+			}
+		}
+		return err
+	})
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			Debugf("Error occurred walking sub directories %q: %s", path, err)
+		}
+	}
+	return dirs
+}
+
+func SystemUserDirsExist(path string) bool {
+	listDirUserPaths := strings.Split(path, string(os.PathSeparator))
+	return len(listDirUserPaths) > SystemUserDirLevel
+}
+
+func SystemUserLevelDirNameGet(path string) string {
```

Comment:
I know your code first checks that it's under `users` and calls `SystemUserDirsExist` which guaranties that `listDirUserPaths` is long enough. But, functions (and their usage) tend to take a life of their own.
So, I think you should either not have them as separate methods or make them more tight in terms of what they check